lib/common/schemas.c

```diff
@@ -1331,3 +1331,115 @@ pcmk__schemas_later_than(const char *name)
 
     return lst;
 }
+
+static void
+external_refs_in_schema(GList **list, const char *contents)
+{
+    xmlNode *ptr = string2xml(contents);
+    /* local-name()= is needed to ignore the xmlns= setting at the top of
+     * the XML file.  Otherwise, the xpath query will always return nothing.
+     */
+    xmlXPathObjectPtr xpathObj = xpath_search(ptr, "//*[local-name()='externalRef'] | //*[local-name()='include']");
+
+    for (int i = 0; i < numXpathResults(xpathObj); i++) {
+        xmlNode *result = getXpathResult(xpathObj, i);
+        const char *href = NULL;
+
+        if (result == NULL) {
+            continue;
+        }
+
+        href = crm_element_value(result, "href");
+
+        if (href != NULL) {
+            char *s = strdup(href);
+
+            CRM_ASSERT(s != NULL);
+            *list = g_list_append(*list, s);
+        }
+    }
+
+    freeXpathObject(xpathObj);
+    free_xml(ptr);
+}
+
+static void
+add_schema_file_to_xml(xmlNode *parent, const char *name, GList **already_included)
+{
+    char *contents = NULL;
+    char *path = NULL;
+    xmlNode *file_node = NULL;
+    GList *includes = NULL;
+
+    /* If we already included this file, don't do so again. */
+    if (g_list_find_custom(*already_included, name, (GCompareFunc) strcmp) != NULL) {
+        return;
+    }
+
+    contents = pcmk__schema_contents(name);
+
+    if (contents == NULL) {
+        return;
+    }
+
+    /* Create a new <file path="..."> node with the contents of the file
+     * as a CDATA block underneath it.
+     */
+    file_node = create_xml_node(parent, PCMK__XA_FILE);
+
+    if (file_node == NULL) {
+        free(contents);
+        return;
+    }
+
+    if (pcmk__ends_with(name, ".rng")) {
+        path = strdup(name);
+        CRM_ASSERT(path != NULL);
+    } else {
+        path = crm_strdup_printf("%s.rng", name);
+    }
+
+    crm_xml_add(file_node, PCMK__XA_PATH, path);
+    *already_included = g_list_prepend(*already_included, path);
+
+    xmlAddChild(file_node, xmlNewCDataBlock(parent->doc, (pcmkXmlStr) contents,
+                                            strlen(contents)));
+
+    /* Scan the file for any <externalRef> or <include> nodes and build up
+     * a list of the files they reference.
+     */
+    external_refs_in_schema(&includes, contents);
+
+    /* For each referenced file, recurse to add it (and potentially anything it
+     * references, ...) to the XML.
+     */
+    for (GList *iter = includes; iter != NULL; iter = iter->next) {
+        add_schema_file_to_xml(parent, iter->data, already_included);
+    }
```

Comment:
I could also change this to use `g_list_foreach`, which would require changing the definition of `add_schema_file_to_xml` and adding a little struct somewhere for the user data.  I dunno if it's more readable or not, but I kind of like being in the habit of using the foreach functions.  Meh, I could go either way with this.

Reply:
After playing around with this, I think it's less readable to use g_list_foreach.